python/langsmith/client.py

```diff
@@ -3163,24 +3163,63 @@ def _tracing_thread_handle_batch(
             tracing_queue.task_done()
 
 
-def _tracing_thread_func(client_ref: weakref.ref[Client]) -> None:
+def _tracing_control_thread_func(client_ref: weakref.ref[Client]) -> None:
     client = client_ref()
     if client is None:
         return
     tracing_queue = client.tracing_queue
     assert tracing_queue is not None
 
+    sub_threads: List[threading.Thread] = []
+
     # loop until
     while (
         # the main thread dies
         threading.main_thread().is_alive()
         # or we're the only remaining reference to the client
-        and sys.getrefcount(client) > 3
+        and sys.getrefcount(client) > 3 + len(sub_threads)
         # 1 for this func, 1 for getrefcount, 1 for _get_data_type_cached
     ):
-        if next_batch := _tracing_thread_drain_queue(tracing_queue, 100):
+        print("im looping", tracing_queue.qsize())
+        for thread in sub_threads:
+            if not thread.is_alive():
+                sub_threads.remove(thread)
+        if tracing_queue.qsize() > 1000:
+            new_thread = threading.Thread(
+                target=_tracing_sub_thread_func, args=(weakref.ref(client),)
+            )
+            sub_threads.append(new_thread)
+            new_thread.start()
+        if next_batch := _tracing_thread_drain_queue(tracing_queue):
```

Comment:
How does this get assigned to the right thread?

Reply:
not sure I understand the comment, this function runs all in the same thread, so any thing in this function is happening in this one thread

Reply:
ah! So the main thread also drains the queue and handles requests alongside the subthreads, which it spawns to do the same thing? 

It's weird we have a main thread and a subthread doing the same action, but in different places, but I guess that works

Reply:
yea because I think most uses of this don't need the other threads, can change if we think its better

Reply:
so theoretically could you round robin dequeue records from all the different threads as they're accumulated? Nothing wrong with that, just thinking through how this works

Reply:
there is a single queue shared by all consumer threads. increasing nr of threads just increases the rate at which we drawdown the single queue